src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala

```diff
@@ -543,3 +450,277 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
+
+      case '[' =>
+        visitInternalNameOrArrayReference(desc)
+
+      case _ => // skip over primitive types
+    }
+
+    def visitConstant(const: AnyRef): Unit = const match {
+      case t: Type => visitDescriptor(t.getDescriptor)
+      case _ =>
+    }
+
+    // in principle we could references to annotation types, as they only end up as strings in the
+    // constant pool, not as class references. however, the java compiler still includes nested
+    // annotation classes in the innerClass table, so we do the same. explained in detail in the
+    // large comment in class BTypes.
+    def visitAnnotation(annot: AnnotationNode): Unit = {
+      visitDescriptor(annot.desc)
+      if (annot.values != null) annot.values.asScala foreach visitConstant
+    }
+
+    def visitAnnotations(annots: java.util.List[_ <: AnnotationNode]) = if (annots != null) annots.asScala foreach visitAnnotation
+    def visitAnnotationss(annotss: Array[java.util.List[AnnotationNode]]) = if (annotss != null) annotss foreach visitAnnotations
+
+    def visitHandle(handle: Handle): Unit = {
+      visitInternalNameOrArrayReference(handle.getOwner)
+      visitDescriptor(handle.getDesc)
+    }
+  }
+
+  abstract class GenericSignatureVisitor {
+    def visitInternalName(internalName: InternalName): Unit
+
+    def raiseError(msg: String, sig: String, e: Option[Throwable] = None): Unit
+
+    def visitClassSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.classSignature() }
+    }
+
+    def visitMethodSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.methodSignature() }
+    }
+
+    def visitFieldSignature(sig: String): Unit = if (sig != null) {
+      val p = new Parser(sig)
+      p.safely { p.fieldSignature() }
+    }
+
+    private final class Parser(sig: String) {
+      private var index = 0
+      private val end = sig.length
+
+      private val Aborted: Throwable = new NoStackTrace { }
+      private def abort(): Nothing = throw Aborted
+
+      def safely(f: => Unit): Unit = try f catch {
+        case Aborted =>
+        case NonFatal(e) => raiseError(s"Exception thrown during signature parsing", sig, Some(e))
+      }
+
+      private def current = {
+        if (index >= end) {
+          raiseError(s"Out of bounds, $index >= $end", sig)
+          abort() // Don't continue, even if `notifyInvalidSignature` returns
+        }
+        sig.charAt(index)
+      }
+
+      private def accept(c: Char): Unit = {
+        if (current != c) {
+          raiseError(s"Expected $c at $index, found $current", sig)
+          abort()
+        }
+        index += 1
+      }
+
+      private def skip(): Unit = { index += 1 }
+      private def getCurrentAndSkip(): Char = { val c = current; skip(); c }
+
+      private def skipUntil(isDelimiter: Char => Boolean): Unit = {
+        while (!isDelimiter(current)) { index += 1 }
+      }
```

Comment:
Use a SAM type `trait CharBooleanFunction { def apply(c: Char) => Boolean } ` to avoid boxing.

```
scala> ((c: Char) => true).getClass.getInterfaces
res2: Array[Class[_]] = Array(interface scala.Function1, interface scala.Serializable)
```